### ISSUE TITLE
OpenMP reduction segfaults

### ISSUE THREAD

**Issue by jeffhammond** [flang-compiler/flang#56]
OpenMP reductions cause a segfault.  Serializing the loops or removing the line that performs the reduction eliminate the segfault.  Details below.

# Code excerpt

```fortran
  abserr = 0.0
  addit = (0.5*iterations) * (iterations+1)
  !$omp parallel default(none)                                        &
  !$omp&  shared(B)                                                   &
  !$omp&  firstprivate(order,iterations,addit)                        &
  !$omp&  private(i,j,temp)                                           &
  !$omp&  reduction(+:abserr)
  !$omp do collapse(2)
  do j=1,order
    do i=1,order
      temp = ((real(order,REAL64)*real(i-1,REAL64))+real(j-1,REAL64)) &
           * real(iterations+1,REAL64)
      abserr = abserr + abs(B(i,j) - (temp+addit)) ! THIS SEGFAULTS
    enddo
  enddo
  !$omp end do nowait
  !$omp end parallel
```

# Source Files
[bad.txt](https://github.com/flang-compiler/flang/files/1020220/bad.txt)
[fix.txt](https://github.com/flang-compiler/flang/files/1020222/fix.txt)
[good.txt](https://github.com/flang-compiler/flang/files/1020221/good.txt)
```sh
for i in good bad fix ; do mv $i.txt $i.f90 ; done
```
(The original source comes from https://github.com/ParRes/Kernels)

# Compile
```sh
$ for i in good bad fix ; do flang -Mpreprocess -O2 \
-DPRKVERSION="'2.16'" -mp $i.f90 -o $i ; done
```

# Bug
```sh
$ ./bad 10 1000 32
Parallel Research Kernels version       2.16
Fortran OpenMP Matrix transpose: B = A^T
Number of threads    =       16
Matrix order         =     1000
Tile size            =       32
Number of iterations =       10
Segmentation fault (core dumped)
```
# Remove statement that GDB blames for the segfault
```sh
$ ./good 10 1000 32
Parallel Research Kernels version       2.16
Fortran OpenMP Matrix transpose: B = A^T
Number of threads    =       16
Matrix order         =     1000
Tile size            =       32
Number of iterations =       10
Solution validates
Rate (MB/s):  11205.423425 Avg time (s):   0.001428
```
# Serialize the reduction
```sh
$ ./fix 10 1000 32
Parallel Research Kernels version       2.16
Fortran OpenMP Matrix transpose: B = A^T
Number of threads    =       16
Matrix order         =     1000
Tile size            =       32
Number of iterations =       10
Solution validates
Rate (MB/s):   6674.982582 Avg time (s):   0.002397
```

**Comment by sscalpone** [comment]
Addressed in flang-compiler flang commit 28073f59607a640f408474e334035286ce2dbe7f,

**Comment by jeffhammond** [comment]
I confirmed that this is fixed.
